pkg/config/mesh/watcher.go

```diff
@@ -76,29 +81,28 @@ func NewWatcher(fileWatcher filewatcher.FileWatcher, filename string) (Watcher,
 			log.Warnf("failed to read mesh configuration, using default: %v", err)
 			return
 		}
+		w.handleMeshConfig(meshConfig)
+	})
+	return w, nil
+}
 
-		var handlers []func()
-
-		w.mutex.Lock()
-		if !reflect.DeepEqual(meshConfig, w.mesh) {
-			log.Infof("mesh configuration updated to: %s", spew.Sdump(meshConfig))
-			if !reflect.DeepEqual(meshConfig.ConfigSources, w.mesh.ConfigSources) {
-				log.Infof("mesh configuration sources have changed")
-				//TODO Need to re-create or reload initConfigController()
-			}
-
-			// Store the new mesh.
-			atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&w.mesh)), unsafe.Pointer(meshConfig))
-			handlers = append([]func(){}, w.handlers...)
-		}
-		w.mutex.Unlock()
-
-		// Notify the handlers of the change.
-		for _, h := range handlers {
-			h()
+// NewConfigMapWatcher creates a new Watcher for changes to the given ConfigMap.
+func NewConfigMapWatcher(client kube.Client, namespace, name, key string) Watcher {
+	defaultMesh := DefaultMeshConfig()
+	w := &watcher{mesh: &defaultMesh}
+	c := configmapwatcher.NewController(client, namespace, name, func() {
+		meshConfig, err := w.readConfigMap(key)
+		if err != nil {
+			log.Warnf("failed to read mesh config from ConfigMap, using default: %v", err)
```

Comment:
Do you think its possible to distinguish between a real error and the configmap being deleted? If there is a temporary error and we revert to default config it could cause a mesh wide outage.

Reply:
Thank you for catching this. What I'll plan to do is move the errors into the controller, and have it so that

- If the ConfigMap is not found, use the default
- If there's a transient error, the queue will retry and this handler won't be called
- If the ConfigMap exists but isn't formatted properly, use the last known config

Reply:
BTW - the file watching code actually will use the last known state, not the default if the file is removed. I am not sure which is worse, using default or last known state - if we do default we can flip  all config accidentally, but last known state means Istiod's may be out of sync. Removing the configmap should really never happen though, so probably not a big deal

Reply:
Yeah, I feel like it would be expected that deleting the config would cause issues...